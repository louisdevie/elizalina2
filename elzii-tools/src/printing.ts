export class PrintedCode {
  private _lines: string[]

  public constructor(...lines: string[]) {
    this._lines = lines
  }

  public static join(blocks: PrintedCode[]): PrintedCode {
    return new PrintedCode(...blocks.flatMap((block) => block._lines))
  }

  public toString(): string {
    return this._lines.join('\n')
  }

  public indent(indentation: Indentation, depth: number) {
    const padding = indentation.char.repeat(indentation.size * depth)
    this._lines = this._lines.map((line) => padding + line)
  }
}

export interface Indentation {
  readonly char: string
  readonly size: number
}

/**
 * An interface for code pre- or post-processors.
 */
export interface TextProcessor {
  /**
   * Processes raw code.
   * @param code The original code.
   * @return The new code.
   */
  process(code: string): string
}
