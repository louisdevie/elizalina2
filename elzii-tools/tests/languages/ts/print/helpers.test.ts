import {
  escapeDoubleQuotedString,
  escapeSingleQuotedString,
  wrapInBlock,
} from '@module/languages/ts/print/helpers'
import { PrintedCode } from '@module/printing'

test('escapeSingleQuotedString', () => {
  const escaped = escapeSingleQuotedString('\\ \0 \n \r \v \t \b \f \' " abc 123')
  // the single quote should be escaped but not the double one
  expect(escaped).toEqual('\\\\ \\0 \\n \\r \\v \\t \\b \\f \\\' " abc 123')
})

test('escapeDoubleQuotedString', () => {
  const escaped = escapeDoubleQuotedString('\\ \0 \n \r \v \t \b \f \' " abc 123')
  // here it's the opposite, the double quote should be escaped but not the single one
  expect(escaped).toEqual('\\\\ \\0 \\n \\r \\v \\t \\b \\f \' \\" abc 123')
})

test('wrapInBlock', () => {
  expect(wrapInBlock(new PrintedCode()).toString()).toMatchSnapshot()
  expect(wrapInBlock(new PrintedCode('line1', 'line2')).toString()).toMatchSnapshot()
})
