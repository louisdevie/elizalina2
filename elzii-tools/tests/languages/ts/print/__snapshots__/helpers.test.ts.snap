// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wrapInBlock 1`] = `"{ }"`;

exports[`wrapInBlock 2`] = `
"{
  line1
  line2
}"
`;
